### PR TITLE
Move scrim core

### PR DESCRIPTION
**Pull request by ZarkoPernar**

### CODE REVIEW

File: packages/lab/stories/scrim.doc.stories.mdx [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+**Modal experience**  
+The Scrim provides a modal experience by obscuring the underlying content. It also prevents keyboard or mouse interactions from reaching this content until the Scrim and overlaying component are closed by the user or system.
+
+## Default
+
+When triggered, the basic Scrim fills the application window and blocks user interaction with the underlying content, until it’s closed. It is often displayed as the backdrop to another component, as demonstrated in this example.
+
+<Canvas>
+  <Story name="Default" decorators={[withFlexGap]}>
+    <DefaultScrim />
```
[the remark this answers is not in the document]
Comment (ZarkoPernar) [reply packages/lab/stories/scrim.doc.stories.mdx:44]: Didn't know about the `id="xyz"` but sounds like a good idea.

File: packages/core/src/scrim/Scrim.css [hunk 2]
```diff
@@ -1,9 +1,11 @@
+/* Style applied to the root element */
 .uitkScrim {
-  --scrim-background: var(--uitk-overlayable-background-low);
+  --scrim-background: var(--uitk-overlayable-background-medium);
 }
 
-.uitkEmphasisMedium.uitkScrim {
-  --scrim-background: var(--uitk-overlayable-background-medium);
+/* Apply this className to the element to achieve the lighter color variant */
+.uitkEmphasisLow {
```
[the remark this answers is not in the document]
Comment (ZarkoPernar) [reply packages/core/src/scrim/Scrim.css:7]: @libertymayc isn't this the pattern you were using in the "fix css PR"?

File: packages/lab/stories/scrim.doc.stories.mdx [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+# API
+
+```
+import { Scrim } from "@jpmorganchase/uitk-core";
+```
+
+## Props
+
+<ArgsTable of={Scrim} />
```
[the remark this answers is not in the document]
Comment (ZarkoPernar) [reply packages/lab/stories/scrim.doc.stories.mdx:76]: Good catch on the defaults 👍

[the remark this answers is not in the document]
Comment (ZarkoPernar) [reply packages/lab/stories/scrim.doc.stories.mdx:76]: @origami-z As far as the prop name, yeah, I don't love `containerFix`. However, where does the scope fall in regards to api changes, especially major breaking changes, as part of this work?

File: packages/lab/stories/scrim.doc.stories.mdx [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+<Canvas>
+  <Story name="Default" decorators={[withFlexGap]}>
+    <DefaultScrim />
+  </Story>
+</Canvas>
+
+## Closable Scrim
+
+When triggered, the basic Scrim fills the application window and blocks user interaction with the underlying content, until it’s closed. It is often displayed as the backdrop to another component, as demonstrated in this example.
```
[the remark this answers is not in the document]
Comment (ZarkoPernar) [reply packages/lab/stories/scrim.doc.stories.mdx:50]: Yep, swapped them around.

File: packages/core/src/scrim/Scrim.tsx [hunk 5]
```diff
@@ -123,10 +118,7 @@ export interface ScrimProps extends HTMLAttributes<HTMLDivElement> {
 }
```
Comment (alycrys) [inline packages/core/src/scrim/Scrim.tsx:118]:
```
/**
   * By default, an error will be thrown if the focus trap contains no elements in its tab order.
   * With this option you can specify a fallback element to programmatically receive focus if no other tabbable elements are found.
   * For example, you may want a popover's <div> to receive focus if the popover's content includes no tabbable elements.
   */
  fallbackFocusRef?: FocusManagerProps["fallbackFocusRef"];
```
`<div>` needs to be escaped

File: packages/core/src/scrim/Scrim.css (hunk 2 above)
[the remark this answers is not in the document]
Comment (alycrys) [reply packages/core/src/scrim/Scrim.css:7]: Maybe we should also mention in the comment that 'medium' emphasis is the default?

[the remark this answers is not in the document]
Comment (ZarkoPernar) [reply packages/core/src/scrim/Scrim.css:7]: Sounds good @alycrys

File: packages/core/src/scrim/Scrim.tsx (hunk 5 above)
Comment (ZarkoPernar) [reply packages/core/src/scrim/Scrim.tsx:118]: Good catch 👍

**Review by alycrys — APPROVED** [review]

File: packages/lab/stories/scrim.doc.stories.mdx (hunk 3 above)
[the remark this answers is not in the document]
Comment (joshwooding) [reply packages/lab/stories/scrim.doc.stories.mdx]: API changes are fine. Especially renames since they have an easy migration path.

File: packages/core/src/focus-manager/internal/findAllTabbableElements.ts [hunk 6]
```diff
@@ -1,4 +1,4 @@
-import { ownerWindow } from "../../utils";
+import { ownerWindow } from "../../utils/ownerWindow";
```
Comment (joshwooding) [inline packages/core/src/focus-manager/internal/findAllTabbableElements.ts:1]: ?

File: package.json [hunk 7]
```diff
@@ -24,7 +24,7 @@
     "build:ag-grid-theme": "cd packages/ag-grid-theme && yarn build",
     "test": "modular test",
     "test:components": "cypress run --component --browser chrome --headless",
-    "test:components:local": "cypress open --component --browser chrome",
+    "test:components:local": "cypress open --component --browser electron",
```
[the remark this answers is not in the document]
Comment (ZarkoPernar) [reply package.json:27]: I'll ask about it in tomorrows tech meeting

File: packages/core/src/focus-manager/internal/findAllTabbableElements.ts (hunk 6 above)
Comment (ZarkoPernar) [reply packages/core/src/focus-manager/internal/findAllTabbableElements.ts:1]: Ah yes, I auto-imported this before adding the export

File: packages/lab/stories/scrim.doc.stories.mdx (hunk 3 above)
[the remark this answers is not in the document]
Comment (ZarkoPernar) [reply packages/lab/stories/scrim.doc.stories.mdx]: I decided to do prop refactors in a separate PR #348

File: packages/core/src/scrim/Scrim.css (hunk 2 above)
[the remark this answers is not in the document]
Comment (libertymayc) [reply packages/core/src/scrim/Scrim.css:7]: Not for emphasis - the problem was the space between e.g. .uitk-density-high .uitkScrim - but here we are specifically saying scrim is medium emphasis (no space between the classes), so there shouldn't be nesting problems

**Comment by libertymayc** [comment]
Will need a changeset and also maybe change the pr description :)

File: packages/core/src/scrim/Scrim.css (hunk 2 above)
[the remark this answers is not in the document]
Comment (ZarkoPernar) [reply packages/core/src/scrim/Scrim.css:7]: Ah right, that was density not emphasis 🤦 . I'll revert this change.

**Review by alycrys — APPROVED** [review]

**Review by origami-z — APPROVED** [review]

**Review by joshwooding — APPROVED** [review]

File: packages/core/src/scrim/Scrim.tsx [hunk 8]
```diff
@@ -24,13 +18,15 @@ import "./Scrim.css";
...
+const defaultParent = () =>
+  typeof document !== "undefined" ? document.body : null;
 
+// eslint-disable-next-line @typescript-eslint/no-empty-function
 const noop = () => {};
 
 const withBaseName = makePrefixer("uitkScrim");
 
-function preventSelection(parent = defaultParent): () => void {
+function preventSelection(parent = defaultParent()): () => void {
```
[the remark this answers is not in the document]
Comment (ZarkoPernar) [reply packages/core/src/scrim/Scrim.tsx:29]: I wanted this to be evaluated every time the function runs and not just when the module is loaded, like it was before, in case of server side rendering.

[the remark this answers is not in the document]
Comment (EleanorHaproff) [reply packages/core/src/scrim/Scrim.tsx:29]: 👍 makes sense. Maybe should be getDefaultParent then

[the remark this answers is not in the document]
Comment (ZarkoPernar) [reply packages/core/src/scrim/Scrim.tsx:29]: Sounds good

File: package.json (hunk 7 above)
Comment (joshwooding) [inline package.json:27]: Could we create a new PR for this please

File: packages/lab/src/layout/LayerLayout/LayerLayout.tsx [hunk 9]
```diff
@@ -1,12 +1,12 @@
 import { forwardRef, HTMLAttributes, useState, useEffect } from "react";
 import cx from "classnames";
 
+import { Scrim, ScrimProps } from "@jpmorganchase/uitk-core";
```
Comment (joshwooding) [inline packages/lab/src/layout/LayerLayout/LayerLayout.tsx:4]: These imports should merging

File: package.json (hunk 7 above)
Comment (ZarkoPernar) [reply package.json:27]: Ah, yes, forgot about this

File: packages/lab/src/layout/LayerLayout/LayerLayout.tsx (hunk 9 above)
Comment (ZarkoPernar) [reply packages/lab/src/layout/LayerLayout/LayerLayout.tsx:4]: You mean, should not be merged?

Comment (ZarkoPernar) [reply packages/lab/src/layout/LayerLayout/LayerLayout.tsx:4]: Should I do the following?

```ts
import { Scrim } from "@jpmorganchase/uitk-core";
import type { ScrimProps } from "@jpmorganchase/uitk-core";
```

Comment (joshwooding) [reply packages/lab/src/layout/LayerLayout/LayerLayout.tsx:4]: Sorry, as in with the other imports from the core package in the file.

Comment (ZarkoPernar) [reply packages/lab/src/layout/LayerLayout/LayerLayout.tsx:4]: Ah, right. Im blind :'D

File: packages/core/src/__tests__/__e2e__/scrim/Scrim.cy.tsx [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+          </>
+        );
+      }
+      cy.mount(<TestComponent />);
+
+      cy.findByRole("button", { name: "OPEN SCRIM" }).click();
+      cy.findByRole("button", { name: "CLOSE SCRIM" })
+        .should("have.focus")
+        .realClick();
+      cy.wait(50);
```
Comment (joshwooding) [inline packages/core/src/__tests__/__e2e__/scrim/Scrim.cy.tsx:217]: Is the wait actually needed?

**Review by joshwooding — APPROVED** [review]

File: packages/core/src/__tests__/__e2e__/scrim/Scrim.cy.tsx (hunk 10 above)
Comment (ZarkoPernar) [reply packages/core/src/__tests__/__e2e__/scrim/Scrim.cy.tsx:217]: Yeah, this one is a bit hard to test because if you removed the wait it would still pass iirc. But that would be a false positive because the return of the focus takes a bit of time so we don't want to check immediately.

Comment (joshwooding) [reply packages/core/src/__tests__/__e2e__/scrim/Scrim.cy.tsx:217]: Not sure how it would pass as a false positive though?

Comment (ZarkoPernar) [reply packages/core/src/__tests__/__e2e__/scrim/Scrim.cy.tsx:217]: Because what we are trying to test here is the return of focus functionality, and that happens at a delay. If you test immediately, you are not really testing that functionality.

Comment (joshwooding) [reply packages/core/src/__tests__/__e2e__/scrim/Scrim.cy.tsx:217]: But if it returns true the element has focus which is the correct behaviour? Cypress should manage the retry or the delay if needed?

Comment (ZarkoPernar) [reply packages/core/src/__tests__/__e2e__/scrim/Scrim.cy.tsx:217]: We are testing that the previous button does **not** have focus. If we assert that immediately it will return true, because it doesn't, and then cypress will move on but at the time of the assertion the return focus hasn't done its thing yet. There is probably a better way to test this that I didn't think of but just removing the `wait()` is not it. The way I made sure this test was making the right assertions was to toggle the return focus functionality on/off while commenting out the wait command.

Comment (ZarkoPernar) [reply packages/core/src/__tests__/__e2e__/scrim/Scrim.cy.tsx:217]: Tbh, I don't even remember the specifics of that scenario just that I did play around with this and confirmed that I need the wait. I don't like adding arbitrary waits in my tests but usually it is the best way to prove a negative in a test.

Comment (ZarkoPernar) [reply packages/core/src/__tests__/__e2e__/scrim/Scrim.cy.tsx:217]: Unless you have something else in the ui to wait for. Here I don't think I have. Maybe waiting for the Scrim to disappear but I'm not sure of the timings. 🤔

**Review by joshwooding — APPROVED** [review]
